### DISCUSSION TITLE
Flyback Musik with Floppy?!

### DISCUSSION THREAD

**Discussion by HIGHVOLTAGEBEE** [SammyIAm/Moppy2#222] [General]
Hello, 

I´m trying to make and 4 channel Flyback Transformer music rig.
The thing is taht i don´t now how to change the code to NOT limit the playing frequenzy,
 becaus the flybacks can play high notes. Is taht possible?

Can you please help?

**Comment by DJthefirst** [comment]
So that is a limitation with how Moppy is fundamentally designed. The microcontroller you use can only update every 40us or so let's say. As you get higher frequency you need to toggle a pin every 956us for C6 as it is not divisible it becomes out of tune and up to a certain point unplayable. Your best option if you want a midi on fly back would probably be to get an esp32 and use a synth project with an analog jack paired with an analog jack fly back driver circuit. The con being your limited to only one fly back per device but pro being you can generate sin triangle and square wave sounds on your flayback.

**Comment by DJthefirst** [comment]
I on my work in progress project Mechanical Midi Music encountered this same problem and am currently working on a solution. I am using an AD9833 signal gen chip external to my microcontroller to play any frequency up to the 100khz range. Far higher than you can hear. For Moppy you would need someone to implement a custom instrument type for you though as well as some additional circuitry such as op amps to increase voltage to what you can play. Easiest option would be forgo midi and just follow something like Franzolis analog fly back driver. I have a box of flybacks I need to build into an instrument eventually...

**Reply by HIGHVOLTAGEBEE** [reply]
The thing is that i want to use more than 2 flybacks (4 if you are interested) and be able to control each separately to get the same effect as the moppy project. Any Idea? Or do you now some way of controllong 2 stero jacks with 1 pc?

**Reply by DJthefirst** [reply]
You should with multiple ad9833 and Arduino but I haven't fully tested it yet and you would need to write your own instrument class and add an op amp.

**Reply by DJthefirst** [reply]
Or use 4 esp32 audio jack synths and a usb hub. Then you just program each to a different midi channel.

**Reply by HIGHVOLTAGEBEE** [reply]
OK but i tested it i i m good with the frequenzy limit but some times the pin is on even if its not playing any note is that normal?

**Reply by samphonic** [reply]
Been meaning to reply (and might have other thoughts later when I'm on a PC), but I believe the pin staying on may be normal.  The drives step only on the signal transition (rising or falling, I don't remember) so I think the code just stops the toggling regardless of its current state.

**Reply by DJthefirst** [reply]
That is normal behavior. The code only toggles a pin. Add a line of code for note off that sets the pin low. It might have a sharper end of note sound as it cuts off mid oscillation.

**Comment by samphonic** [comment]
> The thing is that i want to use more than 2 flybacks (4 if you are interested) and be able to control each separately to get the same effect as the moppy project. Any Idea? 

It sounds like you've sorta got things working, but the short answer is: basically because the output of the `STEP` pins is a square wave at the frequency of the note, you can just use that as your "audio".  As @DJthefirst said though, there is sort of a fundamental limit on the frequency because of how the [code was written](https://github.com/Sammy1Am/Moppy2/blob/master/Microcontroller/Moppy2-Arduino/src/MoppyInstruments/MoppyInstrument.h) and the clock speed of the Arduino.

> Or do you now some way of controllong 2 stero jacks with 1 pc?

I *really* thought this would be easy, but I'm coming up dry on a Google search.  There are plenty of mixer programs (e.g. Voicemeeter Banana) that will let you do all sorts of audio routing to send audio to different outputs.  And most PCs have multiple stereo jacks (or you could get a USB add on).  **BUT** I can't find any way to split up the MIDI so that different MIDI-channels are synthesized on different audio-channels.  So if you manage to figure that out, this might be the slightly easier way than going via Moppy (you could still even use Moppy for its script mapping capabilities and just route the MIDI Out from Moppy via a virtual MIDI cable to whatever synth).

**Reply by HIGHVOLTAGEBEE** [reply]
Can you edit the code in a way that the pin only is on when the note is playing becaus if its on continuous 1 Flyback is using over 14 Amps but if its playing a note only the its halved. Would be great.

**Reply by samphonic** [reply]
I don't know if I'll have a chance anytime soon, but essentially what you're looking at is adding an additional statement below [this line](https://github.com/Sammy1Am/Moppy2/blob/d5fcab0a1faf911800786e7691eb7963da4b3879/Microcontroller/Moppy2-Arduino/src/MoppyInstruments/FloppyDrives.cpp#L129) to the effect of:
``` c
digitalWrite(subAddress*2, LOW);
```

That will at least set the pin to `LOW` whenever `dev_noteOff` is called.  There might be some edge cases that aren't accounted for where you'd have to add essentially that same line above.

**Reply by DJthefirst** [reply]
Also add the line to ground all pins under system reset and device reset. Also note that shunting the pin off without it being aligned with the notes natural toggling can cause a harsher end to the note.

**Reply by HIGHVOLTAGEBEE** [reply]
Thanks!
